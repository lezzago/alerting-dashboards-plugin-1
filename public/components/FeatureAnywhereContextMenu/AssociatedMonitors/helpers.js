import React, { useMemo } from 'react';
import { EuiHealth } from '@elastic/eui';
import { stateToLabel } from '../../../utils/contextMenu/monitors';
import { dateOptionsLong } from '../../../utils/contextMenu/helpers';

export const useColumns = ({ onUnlink, onEdit }) => {
  const columns = useMemo(
    () => [
      {
        field: 'name',
        name: 'Monitor Name',
        sortable: true,
        truncateText: true,
        width: '50%',
      },
      {
        field: 'state',
        name: 'State',
        sortable: true,
        width: '105px',
        render: (state) => (
          <EuiHealth color={stateToLabel[state].color}>{stateToLabel[state].label}</EuiHealth>
        ),
      },
      {
        field: 'date',
        name: 'Last alert',
        sortable: true,
        truncateText: true,
        width: '50%',
        render: (date) => new Intl.DateTimeFormat('default', dateOptionsLong).format(date),
      },
      {
        name: 'Actions',
        actions: [
          {
            type: 'icon',
            name: 'Edit monitor',
            description: 'Edit monitor',
            icon: 'pencil',
            onClick: onEdit,
          },
          {
            type: 'icon',
            name: 'Unlink monitor',
            description: 'Unlink monitor',
            icon: 'unlink',
            onClick: onUnlink,
          },
        ],
      },
    ],
    [onUnlink, onEdit]
  );
  return columns;
};
