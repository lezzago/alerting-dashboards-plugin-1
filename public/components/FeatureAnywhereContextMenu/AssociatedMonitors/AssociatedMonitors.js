import React, { useCallback, useMemo } from 'react';
import {
  EuiFlyoutHeader,
  EuiTitle,
  EuiText,
  EuiSpacer,
  EuiInMemoryTable,
  EuiFlyoutBody,
  EuiEmptyPrompt,
  EuiButton,
} from '@elastic/eui';
import uuidv4 from 'uuid/v4';
import './styles.scss';
import { getColumns } from './helpers';

const AssociatedMonitors = ({ embeddable, closeFlyout, setPanel }) => {
  const title = embeddable.getTitle();
  const onUnlink = useCallback(
    (item) => {
      console.log('onUnlink', item);
      closeFlyout();
    },
    [closeFlyout]
  );
  const onEdit = useCallback(
    (item) => {
      console.log('onEdit', item);
      closeFlyout();
    },
    [closeFlyout]
  );
  const onView = useCallback(
    (item) => {
      console.log('onView', item);
      closeFlyout();
    },
    [closeFlyout]
  );
  const columns = useMemo(() => getColumns({ onUnlink, onEdit, onView }), [
    onUnlink,
    onEdit,
    onView,
  ]);
  const monitors = [
    // { name: 'CPU usage across world', state: 'enabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 2', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 3', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 4', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 5', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 6', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 7', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 8', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 9', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 10', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 11', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 12', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 13', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 14', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 15', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 16', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 17', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 18', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 19', state: 'disabled', date: Date.now(), id: uuidv4() },
    // { name: 'Memory usage across world 20', state: 'disabled', date: Date.now(), id: uuidv4() },
  ];
  const empty = (
    <EuiEmptyPrompt
      title={<h3>No monitors to display</h3>}
      titleSize="s"
      body={`There are no alerting monitors associated with ${title} visualization. You will need to add a monitor to the visualization to be able to list it here`}
      actions={
        <EuiButton fill onClick={() => setPanel('add')}>
          Add alerting monitor
        </EuiButton>
      }
    />
  );
  const tableProps = {
    items: monitors,
    columns,
    search: {
      box: {
        disabled: monitors.length === 0,
        incremental: true,
        schema: true,
      },
    },
    hasActions: true,
    pagination: true,
    sorting: true,
    message: empty,
  };

  return (
    <div className="associated-monitors">
      <EuiFlyoutHeader hasBorder>
        <EuiTitle size="s">
          <h2 id="associated-monitors__title">
            Associated monitors {monitors.length > 0 ? `(${monitors.length})` : ''}
          </h2>
        </EuiTitle>
      </EuiFlyoutHeader>
      <EuiFlyoutBody>
        <EuiText>
          <h4>{title}</h4>
        </EuiText>
        <EuiSpacer size="l" />
        <EuiInMemoryTable {...tableProps} />
      </EuiFlyoutBody>
    </div>
  );
};

export default AssociatedMonitors;
