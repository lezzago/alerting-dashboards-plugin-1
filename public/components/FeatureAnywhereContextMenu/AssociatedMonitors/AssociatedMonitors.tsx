import React, { useCallback, useState, useEffect } from 'react';
import {
  EuiCallOut,
  EuiFlyout,
  EuiFlyoutHeader,
  EuiTitle,
  EuiText,
  EuiSpacer,
  EuiInMemoryTable,
  EuiFlyoutBody,
  EuiEmptyPrompt,
  EuiButton,
  EuiFlexGroup,
  EuiFlexItem,
  EuiLoadingSpinner,
  EuiTextColor,
} from '@elastic/eui';
import './styles.scss';
import { useColumns } from './helpers';
import { ConfirmUnlinkDetectorModal } from './ConfirmUnlinkModal';
import {
  deleteAlertingAugmentVisSavedObj, getAssociatedMonitorIds,
  getCountOfAssociatedObjects,
  validateAssociationIsAllow,
} from '../../../utils/savedObjectHelper';
import { getClient, getNotifications, getSavedAugmentVisLoader, getUISettings } from '../../../services';
import { PLUGIN_AUGMENTATION_MAX_OBJECTS_SETTING } from '../../../utils/constants';

const AssociatedMonitors = ({ embeddable, closeFlyout, setFlyoutMode, monitors }) => {
  const title = embeddable.vis.title;
  const [modalState, setModalState] = useState(undefined);
  const notifications = getNotifications();
  const onUnlink = useCallback((item) => {
    setModalState({
      monitor: {
        name: item.name,
        id: item.id,
      },
    });
  }, []);
  const onEdit = useCallback(
    (item) => {
      window.open(`alerting#/monitors/${item.id}?action=update-monitor`, '_blank');
      // closeFlyout();
    },
    [closeFlyout]
  );
  const onUnlinkMonitor = useCallback(async () => {
    try {
      await deleteAlertingAugmentVisSavedObj(embeddable.vis.id, modalState.monitor.id);
      notifications.toasts.addSuccess({
        title: `Association removed between the ${modalState.monitor.name} monitor and the ${title} visualization`,
        text:
          "The monitor's alerts do not automatically appear on the visualization. Refresh your dashboard to update the visualization.",
      });
      // Need to remove the monitor the below does not work.
      // maybe need to export
      // setAllMonitors(monitors.filter((monitor) => monitor.id !== modalState.monitor.id));
    } catch (e) {
      notifications.toasts.addDanger(
        `Failed to remove the association between the "${modalState.monitor.name}" monitor with the ${title} visualization. Failed due to ${e.message}.`
      );
    } finally {
      handleHideModal();
    }
  }, [closeFlyout, modalState]);
  const columns = useColumns({ onUnlink, onEdit });
  const emptyMsg = (
    <EuiEmptyPrompt
      title={<h3>No monitors to display</h3>}
      titleSize="s"
      body={`There are no alerting monitors associated with ${title} visualization.`}
    />
  );
  const noResultsMsg = (
    <EuiEmptyPrompt
      title={<h3>No monitors to display</h3>}
      titleSize="s"
      body="There are no alerting monitors that match the search result."
    />
  );

  const loadingMsg = <EuiEmptyPrompt body={<EuiLoadingSpinner size="l" />} />;
  const tableProps = {
    items: monitors || [],
    columns,
    search: {
      box: {
        disabled: !monitors || monitors.length === 0,
        incremental: true,
        schema: true,
      },
    },
    hasActions: true,
    pagination: true,
    sorting: true,
    message: monitors ? (monitors.length ? noResultsMsg : emptyMsg) : loadingMsg,
    loading: !monitors,
  };

  const handleHideModal = useCallback(() => {
    setModalState(undefined);
  }, []);

  //TODO: move this to Containers and pass this in. Also make sure its exportable to be updated
  const [isAssociateAllowed, setIsAssociateAllowed] = useState<any[] | null>();

  useEffect(() => {
    const getIsAssociateAllowed = async () => {
      const isAllowed = await validateAssociationIsAllow(embeddable.vis.id);
      setIsAssociateAllowed(isAllowed);
    };

    getIsAssociateAllowed();
  }, []);

  const uiSettings = getUISettings();
  const maxAssociatedCount = uiSettings.get(PLUGIN_AUGMENTATION_MAX_OBJECTS_SETTING);
  const limitReachedTitle = `Limit reached. No more than ${maxAssociatedCount} objects can be associated with a visualizations.`

  return (
    <div className="associated-monitors">
      {modalState ? (
        <ConfirmUnlinkDetectorModal
          monitor={modalState.monitor}
          onHide={handleHideModal}
          onConfirm={onUnlinkMonitor}
          isListLoading={!monitors}
        />
      ) : null}
      <EuiFlyout ownFocus size="m" paddingSize="m" onClose={closeFlyout}>
        <EuiFlyoutHeader hasBorder>
          <EuiTitle>
            <h2 id="associated-monitors__title">Associated monitors</h2>
          </EuiTitle>
        </EuiFlyoutHeader>
        {!isAssociateAllowed && (
          <EuiFlyoutHeader>
            <EuiCallOut title={limitReachedTitle} color="warning" iconType="alert">
              Adding more objects may affect cluster performance and prevent dashboards from rendering properly. Remove associations before add new ones.
            </EuiCallOut>
          </EuiFlyoutHeader>
        )}
        <EuiFlyoutBody>
          <EuiFlexGroup justifyContent="spaceBetween" alignItems="center">
            <EuiFlexItem>
              <EuiText>
                <h4>Visualization: {title}</h4>
              </EuiText>
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <EuiButton iconType="link" isDisabled={!isAssociateAllowed} fill onClick={() => setFlyoutMode('existing')}>
                Associate a monitor
              </EuiButton>
            </EuiFlexItem>
          </EuiFlexGroup>
          <EuiSpacer size="l" />
          <EuiInMemoryTable {...tableProps} />
        </EuiFlyoutBody>
      </EuiFlyout>
    </div>
  );
};

export default AssociatedMonitors;
