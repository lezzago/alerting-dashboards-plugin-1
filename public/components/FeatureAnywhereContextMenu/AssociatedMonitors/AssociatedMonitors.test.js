/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { shallow } from 'enzyme';
import AssociatedMonitors from './AssociatedMonitors';

describe('AssociatedMonitors', () => {
  test('renders', () => {
    const wrapper = shallow(<AssociatedMonitors {...{ embeddable: { getTitle: () => '' } }} />);
    expect(wrapper).toMatchSnapshot();
  });
});
