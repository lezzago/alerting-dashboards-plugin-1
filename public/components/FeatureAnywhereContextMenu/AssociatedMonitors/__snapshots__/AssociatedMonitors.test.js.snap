// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssociatedMonitors renders 1`] = `
<div
  className="associated-monitors"
>
  <EuiFlyoutHeader
    hasBorder={true}
  >
    <EuiTitle
      size="s"
    >
      <h2
        id="associated-monitors__title"
      >
        Associated monitors 
      </h2>
    </EuiTitle>
  </EuiFlyoutHeader>
  <EuiFlyoutBody>
    <EuiText>
      <h4 />
    </EuiText>
    <EuiSpacer
      size="l"
    />
    <EuiInMemoryTable
      columns={
        Array [
          Object {
            "field": "name",
            "name": "Monitor Name",
            "sortable": true,
            "truncateText": true,
            "width": "50%",
          },
          Object {
            "field": "state",
            "name": "State",
            "render": [Function],
            "sortable": true,
            "width": "105px",
          },
          Object {
            "field": "date",
            "name": "Last alert",
            "render": [Function],
            "sortable": true,
            "truncateText": true,
            "width": "50%",
          },
          Object {
            "actions": Array [
              Object {
                "description": "Unlink monitor",
                "icon": "unlink",
                "name": "Unlink monitor",
                "onClick": [Function],
                "type": "icon",
              },
              Object {
                "description": "Edit monitor",
                "icon": "pencil",
                "name": "Edit monitor",
                "onClick": [Function],
                "type": "icon",
              },
              Object {
                "description": "View configuration",
                "icon": "popout",
                "name": "View configuration",
                "onClick": [Function],
                "type": "icon",
              },
            ],
            "name": "Actions",
          },
        ]
      }
      hasActions={true}
      items={Array []}
      message={
        <EuiEmptyPrompt
          actions={
            <EuiButton
              fill={true}
              onClick={[Function]}
            >
              Add alerting monitor
            </EuiButton>
          }
          body="There are no alerting monitors associated with  visualization. You will need to add a monitor to the visualization to be able to list it here"
          title={
            <h3>
              No monitors to display
            </h3>
          }
          titleSize="s"
        />
      }
      pagination={true}
      responsive={true}
      search={
        Object {
          "box": Object {
            "disabled": true,
            "incremental": true,
            "schema": true,
          },
        }
      }
      sorting={true}
      tableLayout="fixed"
    />
  </EuiFlyoutBody>
</div>
`;
